deepspeed/runtime/zero/partition_parameters.py

```diff
@@ -548,6 +750,14 @@ def _post_init_method(self, module):
                 print_rank_0(
                     f"Partitioning param {debug_param2name_id_shape(param)} module={debug_module2name(module)}"
                 )
+
+                if param.is_cuda:
+                    torch.distributed.broadcast(param, 0, self.ds_process_group)
```

Comment:
Why is broadcast needed here? Did you discover a scenario where the normal partition/gather sequence is not sufficient to handle?

Reply:
We need to make sure each rank receives the respective chunk from the same weight tensor. This is important if some pytorch module uses the tensor based initialization method such as orthogonal initialization. This is also why we first gather tensor before apply the user defined init function in https://github.com/jfc4050/DeepSpeed/blob/s3-pr/deepspeed/runtime/zero/partition_parameters.py#L331-L332. It also helps unit-test for the correctness.

Reply:
@szhengac, thanks for the explanation. I am not familiar with this orthogonal initialization approach, but that is not a big deal. What is not clear to me is the following: 
1. Will the all-gather of a given tensor happen on all ranks?
2. If so, then won't the tensor value be the same on all the ranks?
3. If so, what will the broadcast achieve?

Can you please clarify which of the 3 steps I am getting wrong? Thanks.

Reply:
1. yes
2. no, because after the allgather we perform some initialization operation that will give different results on each rank unless seeding is carefully controlled
3. the broadcast ensures that rank 0 is used as the "master" set of initialized weights - this is important if elements are not initialized independently - a simple example is lets say we have an initialization operation that results in the parameter having norm 1, two ranks may individually initialize in a way that satisfies this property, but if you had each rank partition and then did an allgather the resulting tensor is not guaranteed to satisfy this property

Reply:
Got it. We have always assumed seeding is carefully controlled, which is of course is not full-proof :). This looks like a bug fix to me, thanks!

Reply:
yw! i'll add some comments here, if you found it unintuitive i bet others will as well